runtime/vm/exceptiondescribe.c

```diff
@@ -154,62 +151,45 @@ printStackTraceEntry(J9VMThread * vmThread, void * voidUserData, UDATA bytecodeO
 				}
 			}
 		}
-		if (romMethod->modifiers & J9AccNative) {
+		if (J9_ARE_ANY_BITS_SET(romMethod->modifiers, J9AccNative)) {
 			sourceFileName = "NativeMethod";
 			sourceFileNameLen = LITERAL_STRLEN("NativeMethod");
-		} else if (sourceFile) {
+		} else if (NULL != sourceFile) {
 			sourceFileName = (char *)J9UTF8_DATA(sourceFile);
 			sourceFileNameLen = (UDATA)J9UTF8_LENGTH(sourceFile);
 		} else {
 			sourceFileName = "Unknown Source";
 			sourceFileNameLen = LITERAL_STRLEN("Unknown Source");
 		}
 		if (NULL != moduleNameUTF) {
-			if (NULL != moduleVersionUTF) {
-				if (lineNumber) {
-					format = j9nls_lookup_message(
-						J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
-						J9NLS_VM_STACK_TRACE_WITH_MODULE_VERSION_LINE,
-						"\tat %.*s.%.*s (%s@%s/%.*s:%u)\n");
-				} else {
-					format = j9nls_lookup_message(
-						J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
-						J9NLS_VM_STACK_TRACE_WITH_MODULE_VERSION,
-						"\tat %.*s.%.*s (%s@%s/%.*s)\n");
-				}
-				j9tty_err_printf(PORTLIB, (char*)format,
-					(UDATA)J9UTF8_LENGTH(className), J9UTF8_DATA(className),
-					(UDATA)J9UTF8_LENGTH(methodName), J9UTF8_DATA(methodName),
-					moduleNameUTF, moduleVersionUTF,
-					sourceFileNameLen, sourceFileName,
-					lineNumber); /* line number will be ignored in if it's not used in the format string */
-				if (TRUE == freeModuleVersion) {
-					j9mem_free_memory(moduleVersionUTF);
-				}
+			if (0 != lineNumber) {
+				format = j9nls_lookup_message(
+					J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
+					J9NLS_VM_STACK_TRACE_WITH_MODULE_VERSION_LINE,
+					"\tat %.*s.%.*s (%s@%s/%.*s:%u)\n");
 			} else {
-				if (lineNumber) {
-					format = j9nls_lookup_message(
-						J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
-						J9NLS_VM_STACK_TRACE_WITH_MODULE_LINE,
-						"\tat %.*s.%.*s (%s/%.*s:%u)\n");
-				} else {
-					format = j9nls_lookup_message(
-						J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
-						J9NLS_VM_STACK_TRACE_WITH_MODULE,
-						"\tat %.*s.%.*s (%s/%.*s)\n");
-				}
-				j9tty_err_printf(PORTLIB, (char*)format,
-					(UDATA)J9UTF8_LENGTH(className), J9UTF8_DATA(className),
-					(UDATA)J9UTF8_LENGTH(methodName), J9UTF8_DATA(methodName),
-					moduleNameUTF,
-					sourceFileNameLen, sourceFileName,
-					lineNumber); /* line number will be ignored in if it's not used in the format string */
+				format = j9nls_lookup_message(
+					J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
+					J9NLS_VM_STACK_TRACE_WITH_MODULE_VERSION,
+					"\tat %.*s.%.*s (%s@%s/%.*s)\n");
+			}
+			if (NULL == moduleVersionUTF) {
+				moduleVersionUTF = J9_STR(JAVA_SPEC_VERSION);
+			}
```

Comment:
Does this code handle unnamed modules correctly?  I'm finding the diff a little hard to track

Reply:
I'm hoping that testing would catch this if it's wrong.

Reply:
The code in question `moduleVersionUTF = J9_STR(JAVA_SPEC_VERSION)` only affects early booting stage when `java.base` module hasn't been created yet.
Afterwards, the module version is derived from `module->version`.

Reply:
Which says the window for getting it wrong is small (until java.base is created) but will still affect agents that boot early.  I doubt we have tests for this early condition

Reply:
`moduleVersionUTF` related code is under `if (NULL != moduleNameUTF)` which assumes a named module.

Reply:
https://github.com/eclipse-openj9/openj9/blob/dcf40227cde9c4e744b0c9302d3d605b74c4b65e/runtime/vm/exceptiondescribe.c#L127-L128

If we're java11+, `moduleNameUTF` will never be null.

https://github.com/eclipse-openj9/openj9/blob/dcf40227cde9c4e744b0c9302d3d605b74c4b65e/runtime/vm/exceptiondescribe.c#L135

And assuming `findModuleForPackage` returns null for the unnamed module, then we'll treat all class path entries as being in `JAVA_BASE_MODULE` as we'll never change `moduleNameUTF` to be null

Reply:
https://github.com/eclipse-openj9/openj9/blob/dcf40227cde9c4e744b0c9302d3d605b74c4b65e/runtime/vm/exceptiondescribe.c#L164-L178

And all of this code will see `moduleNameUTF == JAVA_BASE_MODULE` and print as though it's in a module

Reply:
Okay, I see that's an existing problem, will set `moduleNameUTF` to `NULL` when the `findModuleForPackage ` returns `NULL` which means an unnamed module.
I think before `java.base` is created, it can assume bootstrap code instead of unnamed module code.

Reply:
Fixed the case for unnamed module.